Components: Reimplement themes-list grid using react-virtualized

ice9js opened the pull request:
This pull request aims to improve the performance of the `themes-list` component used on `themes-showcase` by reimplementing it using `react-virtualized`.

In order to replace the functionality previously provided by `InfiniteScroll`, the component uses `InfiniteLoader`, `WindowScroller`, `AutoSizer` and `Grid` components from `react-virtualized`.

---

To do:

- [x] Show loading placeholders when scrolling fast
- [x] Don't show loading placeholders when there are no more items to be loaded
- [x] Ensure `WindowScroller` is compatible with the variable height header once #11903 is merged
- [x] Fix the issue with items sometimes not loading - the more columns, the more often this comes up
- [x] Ensure the component is compatible with SSR
- [x] Fix issues with multiple lists under each other

---

Testing:

Go to `/themes` and check the following:

- The page loads the initial 20 items with JavaScript disabled.
- Scrolling is 'smooth'.
*You shouldn't ever get 'stuck' once you start scrolling.*
- Two `ThemesList` components render correctly on `/themes` for a Jetpack site.
- The results are reloaded after adding filters or search terms.

ockham commented:
Without having looked at the code much, seems to be working quite nicely! 🎉 


Be sure to also set up a Jetpack site to test this, since this makes for a more complicated test case: two `ThemesList`s instead of just one! (The upper one only contains locally installed themes and is thus generally shorter than the regular WP.com one.)

ice9js commented:
To prevent any confusion - as far as `InfiniteLoader` is concerned, each `Theme` is a separate row.   For the `Grid` component, rows mean actual grid rows.

This allows us to fetch any number of themes at a time, not depending on the actual layout or screen size.

ice9js commented:
The scrolling issues on `single-site-jetpack` were caused by the `WindowScroller` of the second list not recalculating its position after the first list has loaded.  
Though it is not the case here yet, any other DOM change that affects the list's vertical positioning would cause the issue as well.

A solution recommended in [the official docs](https://github.com/bvaughn/react-virtualized/blob/master/docs/WindowScroller.md#updateposition) it to call `WindowScroller.updatePosition` manually.

The approach I took was to register a reference to the list in the closest parent shared with the element that's changing height. The element with changing height should also accept a callback to trigger when such situation occurs.  
By passing this callback all the way from the shared parent, we can use it to trigger `WindowScroller.updatePosition` in a list that's in a different DOM branch than the resizing element.

This seems as a fairly extendable approach, as we can hook into the chain at any moment to update multiple lists, or if we have multiple resizing children. But it does seem quite an anti-pattern.

ice9js commented:
As suggested by @budzanowski, I refactored my previous solution to utilize `WindowScroller.onScroll` callbacks.

The issue with offset grid cells only arises once the user starts scrolling, so we can call `WindowScroller.updatePosition` 'just-in-time' in the `WindowScroller.onScroll` callback. This triggers the update for all `WindowScroller` components connected to the same scroll element ( `window` in this case ).

To prevent performance issues that might come from calling `updatePosition` on every scroll event, I used `lodash.debounce` to call it on the initial event and not more than once a second.

ice9js commented:
Because `react-virtualized` only works with absolute positioning, it's impossible to get it to render anything on the server without explicitly specifying an element's width and height.  
In order to solve this problem, I decided to write our own `FlexboxGrid` component that would still be compatible with `react-virtualized`, but give us better results with SSR.

The trick is to always render the component as a regular block on the first pass, using all the available data at the time, and switch to `position: absolute;` once we know its width.  
Also, instead of positioning each cell separately, we position the viewport as a whole. This allows us to still use CSS for cell positioning within the visible area which gives better results than doing so with JavaScript.

_Sidenote: The initial idea was to use spacer elements instead of absolute positioning, but changing the height of a not absolutely positioned element changes the scroll position and results in an endless scrolling loop._

ice9js commented:
Loding illustration on a big screen:

![showcase-js-enabled](https://cloud.githubusercontent.com/assets/8056203/25071492/9c24735c-22b9-11e7-87bb-2a05933de9be.gif)

folletto commented:
I tested it in Chrome and I found it working well, but I found two issues:

1. If I jump straight down at a later point in the list, it doesn't render. It works only if I proceed scrolling top-to-bottom and letting each block load.
2. At the end I keep getting loading blocks even if the list seems complete. This doesn't happen often, but seems happening reliably if I change site.

Also, I haven't checked performance. In theory there should be an improvement, have we checked?

ice9js commented:
> If I jump straight down at a later point in the list, it doesn't render. It works only if I proceed scrolling top-to-bottom and letting each block load.

The issue here is that `InfiniteLoader` caches the calls to `loadMoreRows` callback, so it doesn't download the same records multiple times, because ideally only the visible items should be downloaded.  
Since we use 'pagination' the approach where we 'render' all placeholders as soon as we know the total number of results won't work, so I went back to progressively appending rows of results as it was done previously.

> At the end I keep getting loading blocks even if the list seems complete. This doesn't happen often, but seems happening reliably if I change site.

If seems that because of SSR, FlexboxGrid was sometimes 'skipping' a page of results.  
I managed to fix this issue by disabling any requests before the component is measured, and disabling the cache for the first request afterwards.

folletto commented:
I really hoped we would be able to render all the placeholder, so the scrollbar was accurate without jumping. But it's ok if we can't work around that, it's more important to not have a regression here.

I did a quick check and it seems working as expected now. I was however able to reproduce again the "doesn't load at the end" if I scroll a bit, switch sites a couple of times (site 1 → site 2 → all sites), and jump at the end again. It's more of an edge case now, but it still happens.

ice9js commented:
I chose to remove the callback cache from `FlexboxGrid` completely, and introduced middleware for `InfiniteList.onRowsRendered()` in `ThemesList` so `InfiniteLoader` keeps track of cells that are actually being loaded instead of the visible ones.  
This allowed me to render all the placeholders at once again and should also fix the issue with overflowing placeholders at the end.

I also fixed a bug where `ThemesList` wouldn't add new cells after being resized - causing the last row to stretch in some cases.

folletto commented:
Tested. Seems FAR more solid, and scrolling feels really good!

I however managed to get the extra placeholders at the end. To reproduce:

1. Open the "All Sites" view of `/themes`
2. Scroll to the very end
3. Switch to a single .com site
4. Scroll to the very end 
5. → placeholders are showing up

ice9js commented:
Thanks for the steps for reproduction. Unfortunately, that issue is not related to `ThemesList` and appears on `master` as well 😮 .

If you repeat all the steps mentioned above on `master`, you'll see that the second *'page'* of results is missing on the same view.  
That's causing the issue here as the amount of themes displayed is lower than the expected `themesCount`.

I'm not sure what it is that's causing the issue yet, but I will make sure to create a separate PR for the fix.

folletto commented:
> Unfortunately, that issue is not related to ThemesList and appears on master as well 😮 .

Nice find :) 
Well not "nice nice" but you get what I mean :P 

I notice that is actually worse in production as it "blinks" visible and not visible, so while jarring this PR is still a slight improvement.

So: let's deal with that separately then as it's not a regression. Let's move forward. :)

ice9js commented:
I opened a separate #13273 for just the `FlexboxGrid` component.

seear commented on client/components/themes-list/index.jsx:
```diff
@@ -109,11 +108,123 @@ export const ThemesList = React.createClass( {
+			) {
+				onRowsRendered( {
+					startIndex: Math.min( this.props.themes.length, stopIndex ),
+					stopIndex: Math.min( this.props.themes.length + 19, this.props.themesCount - 1, stopIndex )
```
Would be good to have some explanation of at least the more 'interesting' numbers here, for example 19, 74.

ice9js replied:
I added a self-explanatory constant for `19` and a more elaborate explanation for what's happening in the `rowHeight` calculation.

alisterscott commented:
Closing as this PR is stale - please reopen if necessary
